sw/airborne/modules/ctc/ctc.c

```diff
@@ -0,0 +1,285 @@
+    float distance_target_ref = sqrtf(error_target_ref_x * error_target_ref_x + error_target_ref_y * error_target_ref_y);
+    if (distance_target_ref < 0.1) {
+      distance_target_ref = 0.1;
+    }
+    float aux = (1 - expf(-ctc_control.alpha * distance_target_ref)) / distance_target_ref;
+    float v_ref_x = ctc_control.target_vx + aux * (error_target_x);
+    float v_ref_y = ctc_control.target_vy + aux * (error_target_y);
+
+    ctc_control.ref_px += v_ref_x * dt;
+    ctc_control.ref_py += v_ref_y * dt;
+
+    float error_v_x = ctc_control.v_centroid_x - v_ref_x;
+    float error_v_y = ctc_control.v_centroid_y - v_ref_y;
+    u_vel = -ctc_control.k1 * (-error_v_x * vy + error_v_y * vx);
+
+    float error_ref_x = px - ctc_control.ref_px;
+    float error_ref_y = py - ctc_control.ref_py;
+    u_spa = ctc_control.omega * (1 + ctc_control.k2 * (error_ref_x * vx + error_ref_y * vy));
+  }
+
+  float u = u_vel + u_spa;
+
+  if (autopilot_get_mode() == AP_MODE_AUTO2) {
+    h_ctl_roll_setpoint =
+      -atanf(u * (sqrtf(vx * vx + vy * vy)) / 9.8 / cosf(att->theta));
+    BoundAbs(h_ctl_roll_setpoint, h_ctl_roll_max_setpoint);
+
+    lateral_mode = LATERAL_MODE_ROLL;
+  }
+
+  if (((now - last_transmision) > ctc_control.time_broad) && (autopilot_get_mode() == AP_MODE_AUTO2)) {
+    ctc_send_info_to_nei();
+    last_transmision = now;
+  }
+}
+
+void ctc_send_info_to_nei(void)
+{
+  struct pprzlink_msg msg;
+
+  for (int i = 0; i < CTC_MAX_AC; i++)
+    if (tableNei[i][0] != -1) {
+      msg.trans = &(DefaultChannel).trans_tx;
+      msg.dev = &(DefaultDevice).device;
+      msg.sender_id = AC_ID;
+      msg.receiver_id = tableNei[i][0];
+      msg.component_id = 0;
+      pprzlink_msg_send_CTC_INFO_TO_NEI(&msg, &ctc_control.vx, &ctc_control.vy, &ctc_control.px, &ctc_control.py);
+    }
+}
+
+void parse_ctc_RegTable(void)
+{
+  uint8_t ac_id = DL_CTC_REG_TABLE_ac_id(dl_buffer);
+  if (ac_id == AC_ID) {
+    uint8_t nei_id = DL_CTC_REG_TABLE_nei_id(dl_buffer);
+    for (int i = 0; i < CTC_MAX_AC; i++)
+      if (tableNei[i][0] == -1) {
+        tableNei[i][0] = (int16_t)nei_id;
+        return;
+      }
+  }
+}
+
+void parse_ctc_CleanTable(void)
+{
+  uint8_t ac_id = DL_CTC_REG_TABLE_ac_id(dl_buffer);
+  if (ac_id == AC_ID)
+    for (int i = 0; i < CTC_MAX_AC; i++) {
+      tableNei[i][0] = -1;
+    }
+}
+
+void parse_ctc_NeiInfoTable(void)
+{
+  int16_t sender_id = (int16_t)(SenderIdOfPprzMsg(dl_buffer));
+  for (int i = 0; i < CTC_MAX_AC; i++)
+    if (tableNei[i][0] == sender_id) {
+      last_info[i] = get_sys_time_msec();
+      tableNei[i][1] = (int16_t)(DL_CTC_INFO_TO_NEI_vx(dl_buffer) * 100);
+      tableNei[i][2] = (int16_t)(DL_CTC_INFO_TO_NEI_vy(dl_buffer) * 100);
+      tableNei[i][3] = (int16_t)(DL_CTC_INFO_TO_NEI_px(dl_buffer) * 100);
+      tableNei[i][4] = (int16_t)(DL_CTC_INFO_TO_NEI_py(dl_buffer) * 100);
+      break;
+    }
+}
+
+void parse_ctc_TargetInfo(void)
+{
+  float moving_target_px = DL_CTC_INFO_FROM_TARGET_px(dl_buffer);
```

Comment:
these are local variables, you are not changing the global one

Reply:
fixed